{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 279,
   "id": "9bd600e2-2196-478e-a245-6876853353f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import cobra\n",
    "import copy\n",
    "import pandas as pd\n",
    "import pylab as plt\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "from scipy import stats\n",
    "from numpy import array,empty\n",
    "import six\n",
    "from sympy import Add, Mul\n",
    "from sympy.functions.elementary.miscellaneous import Max, Min\n",
    "from statannot import add_stat_annotation\n",
    "import pickle\n",
    "import os.path\n",
    "from sklearn.metrics import mean_squared_error\n",
    "import statsmodels.api as sm\n",
    "from matplotlib_venn import venn3\n",
    "from copy import deepcopy\n",
    "from cobra.sampling import sample\n",
    "from scipy import stats\n",
    "from statsmodels.stats.multitest import multipletests"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f6850cb-2510-476c-ac34-b8f1c5d41d15",
   "metadata": {},
   "source": [
    "# Load model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "id": "485f5cf0-5067-47a6-b644-dfb42a1d60ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "unconstrained_model = cobra.io.read_sbml_model(\"../mouse_metabolic_model/Khodaee2020SciRep/iMM1865.xml\")\n",
    "unconstrained_model.solver = \"cplex\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f33dc35c-ddad-40c4-87bc-649573570303",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Load metabolic genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "id": "30f51139-ccc9-41cb-9a0a-1130d2d5044a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of metabolic genes = 3173\n"
     ]
    }
   ],
   "source": [
    "USE_FULL_MODEL = True\n",
    "if USE_FULL_MODEL:\n",
    "    df_metabolic_gene = pd.read_csv(\"../data_04082022/metabolic_genes_with_annotation.csv\", index_col=0)\n",
    "    metabolic_genes = set(df_metabolic_gene.index)\n",
    "else:\n",
    "    metabolic_genes = set([g.name for g in unconstrained_model.genes])\n",
    "print('number of metabolic genes = %d'%(len(metabolic_genes)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02f888a4-ca3b-4466-8d27-650eb1023bb7",
   "metadata": {},
   "source": [
    "# Load RNASeq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "id": "389d3899-6285-4f79-862a-a0051007cb0f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ccr2_depleted</th>\n",
       "      <th>naive</th>\n",
       "      <th>wildtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Fuca2</th>\n",
       "      <td>2291.960451</td>\n",
       "      <td>642.335020</td>\n",
       "      <td>2658.656619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dad1</th>\n",
       "      <td>329.890056</td>\n",
       "      <td>195.099738</td>\n",
       "      <td>363.801130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hdac3</th>\n",
       "      <td>199.653155</td>\n",
       "      <td>139.265618</td>\n",
       "      <td>231.927564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Plcl1</th>\n",
       "      <td>4.927242</td>\n",
       "      <td>8.306191</td>\n",
       "      <td>9.095310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Trmt1</th>\n",
       "      <td>65.058988</td>\n",
       "      <td>38.822970</td>\n",
       "      <td>68.189098</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ccr2_depleted       naive     wildtype\n",
       "Fuca2    2291.960451  642.335020  2658.656619\n",
       "Dad1      329.890056  195.099738   363.801130\n",
       "Hdac3     199.653155  139.265618   231.927564\n",
       "Plcl1       4.927242    8.306191     9.095310\n",
       "Trmt1      65.058988   38.822970    68.189098"
      ]
     },
     "execution_count": 254,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_rnaseq1 = pd.read_csv(\"../data_04082022/2_deseq2_ccr2dep_vs_naive_padj0.05.res.csv\", index_col=0)[['ccr2_depleted','naive']]\n",
    "df_rnaseq2 = pd.read_csv(\"../data_04082022/3_deseq2_ccr2dep_vs_wt_padj0.05.res.csv\", index_col=0)[['wildtype']]\n",
    "df_rnaseq = pd.merge(df_rnaseq1, df_rnaseq2, left_index=True, right_index=True, how='inner')\n",
    "df_rnaseq = df_rnaseq.loc[set(df_rnaseq.index).intersection(metabolic_genes)]\n",
    "df_rnaseq.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fd5fc76-de3d-48e5-8f4d-948b8dd33bd8",
   "metadata": {},
   "source": [
    "## use mean +/- std/2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "id": "9c44ae74-a618-452f-99d1-09f4c8b95483",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(-882.4206276775346, 1792.582287014847)"
      ]
     },
     "execution_count": 255,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(np.mean(df_rnaseq['naive'])-np.std(df_rnaseq['naive'])/2, np.mean(df_rnaseq['naive'])+np.std(df_rnaseq['naive'])/2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "id": "55770150-91d9-4e2d-a796-45d78084133e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(-1491.2575359562416, 2605.342801379842)"
      ]
     },
     "execution_count": 256,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(np.mean(df_rnaseq['wildtype'])-np.std(df_rnaseq['wildtype'])/2, np.mean(df_rnaseq['wildtype'])+np.std(df_rnaseq['wildtype'])/2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "id": "fc9e9434-ed95-47a9-a5db-29ced6916c7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(-1690.6848221507603, 2723.7670578999764)"
      ]
     },
     "execution_count": 257,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(np.mean(df_rnaseq['ccr2_depleted'])-np.std(df_rnaseq['ccr2_depleted'])/2, np.mean(df_rnaseq['ccr2_depleted'])+np.std(df_rnaseq['ccr2_depleted'])/2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f51e4f20-7d6c-4d31-9fe1-290e203bd547",
   "metadata": {},
   "source": [
    "## use 25% and 75% quantile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "id": "07ebfe30-957a-47a6-933c-f9bdfd2e7752",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7.291537942000001, 220.0764992)"
      ]
     },
     "execution_count": 258,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(np.quantile(df_rnaseq['naive'], 0.25, axis=0), np.quantile(df_rnaseq['naive'], 0.75, axis=0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 259,
   "id": "e7ea1ccf-1a96-4467-8f32-a0238943d0ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3.8439575039999996, 268.88582210000004)"
      ]
     },
     "execution_count": 259,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(np.quantile(df_rnaseq['ccr2_depleted'], 0.25, axis=0), np.quantile(df_rnaseq['ccr2_depleted'], 0.75, axis=0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "id": "6b3f9c43-a4b3-494e-954d-b0d5ac9a0ecb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2.915363666, 278.4090173)"
      ]
     },
     "execution_count": 260,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(np.quantile(df_rnaseq['wildtype'], 0.25, axis=0), np.quantile(df_rnaseq['wildtype'], 0.75, axis=0))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4428683-ecef-4c37-b2f1-50a7dfb708ad",
   "metadata": {},
   "source": [
    "## make sure that when naive is 0, both wildtype and ccr2_depleted have low rnaseq count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "id": "57b40226-a17c-44e9-be71-8356936447bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Idi1       2.583671\n",
       "Pnck       2.501887\n",
       "Pglyrp3    1.275161\n",
       "Fabp3      1.193377\n",
       "Pnmt       0.695676\n",
       "Name: wildtype, dtype: float64"
      ]
     },
     "execution_count": 261,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_rnaseq[(df_rnaseq.naive==0) & (df_rnaseq.wildtype>0)].sort_values('wildtype', ascending=False).wildtype.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "id": "ce4b92a0-1bdb-4eea-97f8-0ea9eae14c7f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Idi1        6.383801\n",
       "Fabp3       1.590181\n",
       "Aqp2        1.590181\n",
       "Ccnb1ip1    1.557894\n",
       "Ces3a       1.440416\n",
       "Name: ccr2_depleted, dtype: float64"
      ]
     },
     "execution_count": 262,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_rnaseq[(df_rnaseq.naive==0) & (df_rnaseq.ccr2_depleted>0)].sort_values('ccr2_depleted', ascending=False).ccr2_depleted.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dcbdd8b-8269-45f8-ac7f-b4b052dfaa88",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "dab2a58d-ac2e-4716-9cd4-a693bd7b1fa5",
   "metadata": {},
   "source": [
    "# Constraint input fluxes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "id": "5dcda52f-c66e-4270-932c-7c370aa99772",
   "metadata": {},
   "outputs": [],
   "source": [
    "media = {'EX_glc__D_e'  : -5,          # glucose\n",
    "         'EX_gln__L_e'  : -0.7,        # glutamine\n",
    "         'EX_o2_e'      : -1000,       # oxygen\n",
    "         'EX_pi_e'      : -1000,       # phosphate\n",
    "         'EX_fe3_e'     : -1000,       # fe3+\n",
    "         'EX_zn2_e'     : -1000,       # zn2+\n",
    "         'EX_ca2_e'     : -1000,       # ca2+\n",
    "         'EX_cl_e'      : -1000,       # cl-\n",
    "         'EX_k_e'       : -1000,       # k+\n",
    "         'EX_na1_e'     : -1000,       # na+\n",
    "         'EX_chol_e'    : -1000,       # choline\n",
    "         'EX_na1_e'     : -1000,       # sodium\n",
    "         'EX_hco3_e'    : -1000,       # bicarbonate\n",
    "         'EX_leu__L_e'  : -0.05,       # leucine\n",
    "         'EX_ile__L_e'  : -0.05,       # isoleucine\n",
    "         'EX_pro__L_e'  : -0.05,       # proline\n",
    "         'EX_trp__L_e'  : -0.05,       # tryptophan\n",
    "         'EX_met__L_e'  : -0.05,       # methionine\n",
    "         'EX_cys__L_e'  : -0.05,       # cysteine\n",
    "         'EX_val__L_e'  : -0.05,       # valine\n",
    "         'EX_asp__L_e'  : -0.05,       # aspatate\n",
    "         'EX_asn__L_e'  : -0.05,       # asparagine\n",
    "         'EX_ser__L_e'  : -0.05,       # serine\n",
    "         'EX_thr__L_e'  : -0.05,       # threonine\n",
    "         'EX_ala__L_e'  : -0.05,       # alanine\n",
    "         'EX_arg__L_e'  : -0.05,       # arginine\n",
    "         'EX_gly_e'     : -0.05,       # glycine\n",
    "         'EX_his__L_e'  : -0.05,       # histidine\n",
    "         'EX_lys__L_e'  : -0.05,       # lysine\n",
    "         'EX_phe__L_e'  : -0.05,       # phynaalaine\n",
    "         'EX_glu__L_e'  : -0.05,       # glutamate\n",
    "         'EX_tyr__L_e'  : -0.05        # tyrosine\n",
    "        }\n",
    "\n",
    "ex_rxns = [r for r in unconstrained_model.reactions if r.id.startswith('EX')]\n",
    "for r in ex_rxns:\n",
    "    r.lower_bound = 0\n",
    "    \n",
    "for r in media.keys():\n",
    "    try:\n",
    "        rxn = unconstrained_model.reactions.get_by_id(r)\n",
    "        rxn.lower_bound = media[r]\n",
    "        rxn.upper_bound = 1000\n",
    "    except:\n",
    "        print(\"The reaction {} does not exist in model.\".format(str(r)))\n",
    "        continue;\n",
    "\n",
    "fva = cobra.flux_analysis.variability.flux_variability_analysis(unconstrained_model, unconstrained_model.reactions, loopless=False, fraction_of_optimum=0)\n",
    "for r in unconstrained_model.reactions:\n",
    "    r.lower_bound = fva['minimum'][r.id]\n",
    "    r.upper_bound = fva['maximum'][r.id]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "314fe12f-f647-4c7d-b360-a135a4f46b3a",
   "metadata": {},
   "source": [
    "# iMAT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "id": "4132115b-78ae-4385-b300-401e7a251a5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def imat(model, expression_profile, condition, normalization=or2min_and2max, lowq=0.25, highq=0.75, epsilon=0.1):\n",
    "            \n",
    "    \"\"\"\n",
    "    Parameters\n",
    "    ----------\n",
    "    model: genome-scale model\n",
    "    expression_profile: gene expression profiles\n",
    "    condition: experimental condition (must be a column in the expression profile)\n",
    "    normalization: function to process GPR rules if a reaction is associated with multiple isozymes or one enzyme with multiple subunits\n",
    "    lowq: the quantile to calcualte cutoff value for low expression genes\n",
    "    highq: the quantile to calculate cutoff value for high expression values\n",
    "    epsilon: a positive threshold to determine active flux associated with highly expressed reactions\n",
    "    \"\"\"\n",
    "    \n",
    "    # builds a dict with genes as keys and the expression value for the selected sample.\n",
    "    gene_exp = dict(zip(expression_profile.index, expression_profile[condition]))\n",
    "    low_cutoff = np.quantile(expression_profile[condition], lowq, axis=0)\n",
    "    high_cutoff = np.quantile(expression_profile[condition], highq, axis=0)\n",
    "    \n",
    "    # make a disctionary that mappes reaction id to associated gene expression\n",
    "    gene_id_name_mapping = {g.id:g.name for g in model.genes}\n",
    "    reaction_profile = {}\n",
    "    for r in model.reactions:\n",
    "        # this is an important assumption:\n",
    "        # we only constrain reactions using gene expression data when all associated genes can be found in expression profiles.\n",
    "        if len(r.genes) > 0 and all([gene_id_name_mapping[g.id] in list(expression_profile.index) for g in r.genes]):\n",
    "            _, reaction_profile[r.id] = normalization(r, {g.id: gene_exp[gene_id_name_mapping[g.id]] for g in r.genes})\n",
    "        \n",
    "    y_variables = list()\n",
    "    x_variables = list()\n",
    "    constraints = list()\n",
    "    \n",
    "    # flux variance analysis\n",
    "    fva_res = cobra.flux_analysis.variability.flux_variability_analysis(model, list(reaction_profile.keys()), loopless=False, fraction_of_optimum=0)\n",
    "    \n",
    "    for rid, expression in six.iteritems(reaction_profile):\n",
    "        if expression == 0.0:\n",
    "            # reactions shut down completely\n",
    "            model.reactions.get_by_id(rid).lower_bound = 0\n",
    "            model.reactions.get_by_id(rid).upper_bound = 0\n",
    "        elif expression > high_cutoff:     \n",
    "            reaction = model.reactions.get_by_id(rid)\n",
    "            y_pos = model.solver.interface.Variable(\"y_%s_pos\" % rid, type=\"binary\")\n",
    "            y_neg = model.solver.interface.Variable(\"y_%s_neg\" % rid, type=\"binary\")\n",
    "            y_variables.append([y_neg, y_pos])\n",
    "            pos_constraint = model.solver.interface.Constraint(\n",
    "                reaction.flux_expression + y_pos * (fva_res[\"minimum\"][rid] - epsilon),\n",
    "                lb=fva_res[\"minimum\"][rid], name=\"pos_highly_%s\" % rid)  \n",
    "            neg_constraint = model.solver.interface.Constraint(\n",
    "                reaction.flux_expression + y_neg * (fva_res[\"maximum\"][rid] + epsilon),\n",
    "                ub=fva_res[\"maximum\"][rid], name=\"neg_highly_%s\" % rid)\n",
    "            constraints.extend([pos_constraint, neg_constraint])\n",
    "        elif expression < low_cutoff:\n",
    "            reaction = model.reactions.get_by_id(rid)\n",
    "            x = model.solver.interface.Variable(\"x_%s\" % rid, type=\"binary\")\n",
    "            x_variables.append(x)\n",
    "\n",
    "            pos_constraint = model.solver.interface.Constraint(\n",
    "                (1 - x) * fva_res[\"maximum\"][rid] - reaction.flux_expression,\n",
    "                lb=0, name=\"x_%s_upper\" % rid)\n",
    "            neg_constraint = model.solver.interface.Constraint(\n",
    "                (1 - x) * fva_res[\"minimum\"][rid] - reaction.flux_expression,\n",
    "                ub=0, name=\"x_%s_lower\" % rid)\n",
    "            constraints.extend([pos_constraint, neg_constraint])\n",
    "\n",
    "    len_xy = 0\n",
    "    for variable in x_variables:\n",
    "        model.solver.add(variable)\n",
    "        len_xy += 1\n",
    "\n",
    "    for variables in y_variables:\n",
    "        model.solver.add(variables[0])\n",
    "        model.solver.add(variables[1])\n",
    "        len_xy += 1\n",
    "\n",
    "    for constraint in constraints:\n",
    "        model.solver.add(constraint)\n",
    "\n",
    "    model.objective = model.solver.interface.Objective(Add(*[(y[0] + y[1]) for y in y_variables]) + Add(*x_variables), direction=\"max\")\n",
    "    model.solver.problem.parameters.mip.tolerances.integrality.set(1e-9)\n",
    "    solution = model.optimize()\n",
    "    return model, solution, reaction_profile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "id": "d83b201a-d62e-4fe5-8543-9645a7bf6007",
   "metadata": {},
   "outputs": [],
   "source": [
    "def or2min_and2max(reaction, gene_values):\n",
    "    \n",
    "    \"\"\"\n",
    "    Parameters\n",
    "    ----------\n",
    "    reaction: reaction to be considered\n",
    "    gene_values: a dict with gene ids as keys and their expression levels as values\n",
    "    \"\"\"\n",
    "        \n",
    "    gene_expression_array = reaction.gene_reaction_rule.split(' ')\n",
    "    gene_expression_array = [x for x in gene_expression_array if x != '']\n",
    "    if len(gene_expression_array) > 1:        \n",
    "        gene_expression_array = ['('] + gene_expression_array + [')']\n",
    "        #print(reaction.id, gene_expression_array)\n",
    "\n",
    "        # parse substrings from the innermost parenthesis until neither 'or' or 'and' is leftover\n",
    "        while 'or' in gene_expression_array or 'and' in gene_expression_array:\n",
    "\n",
    "            # find the inner most ( and )\n",
    "            for i in range(len(gene_expression_array)):\n",
    "                if gene_expression_array[i] == ')':\n",
    "                    # should be the first ')' in the array\n",
    "                    right_bracket_index = i\n",
    "                    break\n",
    "            for i in range(len(gene_expression_array)):\n",
    "                if gene_expression_array[i] == '(':\n",
    "                    # should be the last '(' before the first ')'\n",
    "                    left_bracket_index = i\n",
    "                if i==right_bracket_index:\n",
    "                    break\n",
    "\n",
    "            # parse the content of the innermost parenthesis\n",
    "            parenthesis_content = gene_expression_array[left_bracket_index+1:right_bracket_index]\n",
    "\n",
    "            gene_left_behind = ''\n",
    "            if 'and' in ''.join(parenthesis_content):\n",
    "                genes = ''.join(parenthesis_content).split('and')\n",
    "                smallest_value = 1e10\n",
    "                for g in genes:\n",
    "                    current_value = gene_values[g]\n",
    "                    if  current_value < smallest_value:\n",
    "                        gene_left_behind = g\n",
    "                        smallest_value = current_value\n",
    "            elif 'or' in ''.join(parenthesis_content):\n",
    "                genes = ''.join(parenthesis_content).split('or')\n",
    "                largest_value = -0.05\n",
    "                for g in genes:\n",
    "                    current_value = gene_values[g]\n",
    "                    if  current_value > largest_value:\n",
    "                        gene_left_behind = g\n",
    "                        largest_value = current_value\n",
    "            else:\n",
    "                assert len(parenthesis_content)==1\n",
    "                gene_left_behind = parenthesis_content[0]\n",
    "\n",
    "            # replace the entire content in the parenthesis with the gene_left_behind\n",
    "            del gene_expression_array[left_bracket_index:right_bracket_index+1]\n",
    "            gene_expression_array.insert(left_bracket_index, gene_left_behind)\n",
    "                \n",
    "    if len(gene_expression_array)!=1:\n",
    "        raise ValueError(\"%s fails assertion test in or2min_and2max.\"%(reactions.id))\n",
    "    else:\n",
    "        return gene_expression_array[0], gene_values[gene_expression_array[0]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "id": "cf3dda52-d329-4d4b-94ed-2c5474de2cf9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('13081', 5)"
      ]
     },
     "execution_count": 268,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "or2min_and2max(unconstrained_model.reactions.get_by_id('24_25VITD2Hm'), {'13081':5})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "id": "6ab915f0-4a25-4319-ab42-ccc40311dfe6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_metabolic_model(model, iMAT_solution, expression_profile, condition, normalization=or2min_and2max):\n",
    " \n",
    "    \"\"\"\n",
    "    Parameters\n",
    "    ----------\n",
    "    model: genome-scale model\n",
    "    iMAT_solution: iMAT flux solution to constrain reference model\n",
    "    expression_profile: gene expression profiles\n",
    "    condition: condition under consideration\n",
    "    normalization: function to process GPR rules if a reaction is associated with \n",
    "                   multiple isozymes or one enzyme with multiple subunits\n",
    "    \"\"\"\n",
    "\n",
    "    gene_exp = dict(zip(expression_profile.index, expression_profile[condition])) # abosolute value\n",
    "    for r in model.reactions:\n",
    "        if condition != 'Sensitive':\n",
    "            if len(r.genes) > 0 and all([identifier.id in expression_profile.index for identifier in r.genes]):\n",
    "                gid, _ = normalization(r, {g.id: gene_exp.get(g.id) for g in r.genes})\n",
    "                to_constrain = iMAT_solution[r.id] * expression_profile.loc[gid,condition]\n",
    "            else:\n",
    "                to_constrain = iMAT_solution[r.id]\n",
    "        else:\n",
    "            to_constrain = iMAT_solution[r.id]\n",
    "            \n",
    "        if to_constrain == 0.0:\n",
    "            r.lower_bound=to_constrain\n",
    "            r.upper_bound=to_constrain\n",
    "        elif to_constrain < 0.0:\n",
    "            r.lower_bound=to_constrain\n",
    "            \n",
    "            # remove reversibility\n",
    "            if r.upper_bound>0.0:\n",
    "                r.upper_bound=0.0\n",
    "        else:\n",
    "            r.upper_bound=to_constrain\n",
    "            \n",
    "            # remove reversibility\n",
    "            if r.lower_bound<0.0:\n",
    "                r.lower_bound=0.0\n",
    "        \n",
    "    return model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dddd7a0d-0cc1-4969-90be-74d744033fbb",
   "metadata": {},
   "source": [
    "# Generate iMAT model for naive cell type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "id": "b31942a7-0e21-4743-a466-55211297ee7a",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "naive_model, naive_solution, _ = imat(\n",
    "    model=deepcopy(unconstrained_model),\n",
    "    expression_profile=df_rnaseq,\n",
    "    condition=\"naive\",\n",
    "    normalization=or2min_and2max,\n",
    "    lowq=0.25, \n",
    "    highq=0.75, \n",
    "    epsilon=0.1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3a1f52a-d536-45fb-a609-76d253eefe99",
   "metadata": {},
   "source": [
    "# Build metabolic models for naive, WT and ccr2-depleted cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "id": "4b242379-ae1d-466c-b4d3-e371134d9d96",
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_metabolic_model(model, iMAT_solution, expression_profile, condition, normalization=or2min_and2max):\n",
    " \n",
    "    \"\"\"\n",
    "    Parameters\n",
    "    ----------\n",
    "    model: genome-scale model\n",
    "    iMAT_solution: iMAT flux solution to constrain reference model\n",
    "    expression_profile: gene expression profiles\n",
    "    condition: condition under consideration\n",
    "    normalization: function to process GPR rules if a reaction is associated with \n",
    "                   multiple isozymes or one enzyme with multiple subunits\n",
    "    \"\"\"\n",
    "\n",
    "    gene_exp = dict(zip(expression_profile.index, expression_profile[condition])) # abosolute value\n",
    "    gene_id_name_mapping = {g.id:g.name for g in model.genes}\n",
    "    for r in model.reactions:\n",
    "        if len(r.genes) > 0 and all([gene_id_name_mapping[g.id] in list(expression_profile.index) for g in r.genes]):\n",
    "            gid, _ = normalization(r, {g.id: gene_exp[gene_id_name_mapping[g.id]] for g in r.genes})\n",
    "            if expression_profile.loc[gene_id_name_mapping[gid],'naive'] == 0.0:\n",
    "                # when genes are not expressed in the naive cell, they are generally very low in other conditions\n",
    "                # so we assume that they follow the same boundary value\n",
    "                to_constrain = iMAT_solution[r.id]\n",
    "            else:\n",
    "                to_constrain = iMAT_solution[r.id] * expression_profile.loc[gene_id_name_mapping[gid],condition] / expression_profile.loc[gene_id_name_mapping[gid],'naive']\n",
    "        else:\n",
    "            to_constrain = iMAT_solution[r.id]\n",
    "            \n",
    "        if to_constrain == 0.0:\n",
    "            r.lower_bound=to_constrain\n",
    "            r.upper_bound=to_constrain\n",
    "        elif to_constrain < 0.0:\n",
    "            r.lower_bound=to_constrain\n",
    "            if r.upper_bound>0.0:\n",
    "                r.upper_bound=0.0\n",
    "        else:\n",
    "            r.upper_bound=to_constrain\n",
    "            if r.lower_bound<0.0:\n",
    "                r.lower_bound=0.0\n",
    "        \n",
    "    return model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "id": "2d2b6745-af2f-47d0-b2d2-11aacb7dfa1b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "building model for condition: naive\n",
      "sampling solution space for condition-specific model (condition = naive)\n",
      "building model for condition: wildtype\n",
      "sampling solution space for condition-specific model (condition = wildtype)\n",
      "building model for condition: ccr2_depleted\n",
      "sampling solution space for condition-specific model (condition = ccr2_depleted)\n"
     ]
    }
   ],
   "source": [
    "constrained_models = {} # models constrained by gene expression data\n",
    "flux_samples = {} # sampling flux solution space of constrained models\n",
    "\n",
    "conditions=['naive','wildtype','ccr2_depleted']\n",
    "for cond in conditions:\n",
    "         \n",
    "    # construct model\n",
    "    print('building model for condition: %s'%(cond))\n",
    "    model = build_metabolic_model(model=deepcopy(unconstrained_model), \n",
    "                                  iMAT_solution=naive_solution,\n",
    "                                  expression_profile=df_rnaseq,\n",
    "                                  condition=cond)\n",
    "    constrained_models.update({cond: model})\n",
    "    \n",
    "    # sample the flux space\n",
    "    print('sampling solution space for condition-specific model (condition = %s)'%(cond))\n",
    "    s = sample(model, 10000, processes=12)\n",
    "    flux_samples.update({cond : s})\n",
    "\n",
    "# consider save the variables to file so that you can load them next time without running from scratch\n",
    "# write variables:\n",
    "pickle.dump(constrained_models, open(\"constrained_models.p\", \"wb\"))\n",
    "pickle.dump(flux_samples, open(\"flux_samples.p\", \"wb\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1477754-7cb0-48b6-ade5-cb4a72ad0c06",
   "metadata": {},
   "source": [
    "# Identify fluxes significant different between each other"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 273,
   "id": "cfa328d4-d062-4f9c-8f6c-ad1c927cd59a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load variables:\n",
    "constrained_models = pickle.load(open(\"constrained_models.p\",\"rb\"))\n",
    "flux_samples = pickle.load(open(\"flux_samples.p\",\"rb\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 285,
   "id": "ce1cc21e-fe9d-49ea-828c-bf143b913fc7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/ipykernel_launcher.py:14: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  \n",
      "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/ipykernel_launcher.py:14: RuntimeWarning: invalid value encountered in log2\n",
      "  \n",
      "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/ipykernel_launcher.py:14: RuntimeWarning: divide by zero encountered in log2\n",
      "  \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Condition1</th>\n",
       "      <th>Condition2</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>MeanFluxCond1</th>\n",
       "      <th>MeanFluxCond2</th>\n",
       "      <th>Log2MeanFluxRatio</th>\n",
       "      <th>P</th>\n",
       "      <th>Padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>10FTHF7GLUtl</td>\n",
       "      <td>0.047169</td>\n",
       "      <td>0.050554</td>\n",
       "      <td>-0.099989</td>\n",
       "      <td>1.674592e-47</td>\n",
       "      <td>1.776560e-47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>10FTHFtl</td>\n",
       "      <td>-0.047169</td>\n",
       "      <td>-0.050554</td>\n",
       "      <td>-0.099989</td>\n",
       "      <td>1.674592e-47</td>\n",
       "      <td>1.776560e-47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>10FTHFtm</td>\n",
       "      <td>467.427271</td>\n",
       "      <td>108.454557</td>\n",
       "      <td>2.107651</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>2HBO</td>\n",
       "      <td>-0.053233</td>\n",
       "      <td>-0.025287</td>\n",
       "      <td>1.073907</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>2HBt2</td>\n",
       "      <td>-0.062086</td>\n",
       "      <td>-0.031279</td>\n",
       "      <td>0.989073</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Condition1 Condition2      Reaction  MeanFluxCond1  MeanFluxCond2  \\\n",
       "0      naive   wildtype  10FTHF7GLUtl       0.047169       0.050554   \n",
       "1      naive   wildtype      10FTHFtl      -0.047169      -0.050554   \n",
       "2      naive   wildtype      10FTHFtm     467.427271     108.454557   \n",
       "3      naive   wildtype          2HBO      -0.053233      -0.025287   \n",
       "4      naive   wildtype         2HBt2      -0.062086      -0.031279   \n",
       "\n",
       "   Log2MeanFluxRatio             P          Padj  \n",
       "0          -0.099989  1.674592e-47  1.776560e-47  \n",
       "1          -0.099989  1.674592e-47  1.776560e-47  \n",
       "2           2.107651  0.000000e+00  0.000000e+00  \n",
       "3           1.073907  0.000000e+00  0.000000e+00  \n",
       "4           0.989073  0.000000e+00  0.000000e+00  "
      ]
     },
     "execution_count": 285,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paired_conditions = [('naive','wildtype'),('naive','ccr2_depleted'),('wildtype','ccr2_depleted')]\n",
    "flux_res = []\n",
    "for pair in paired_conditions:\n",
    "    cond1 = pair[0]\n",
    "    cond2 = pair[1]\n",
    "    flux_samples1 = flux_samples[cond1]\n",
    "    flux_samples2 = flux_samples[cond2]\n",
    "    for rid in flux_samples1.columns:\n",
    "        x = list(flux_samples1[rid])\n",
    "        y = list(flux_samples2[rid])\n",
    "        if np.max(np.abs(x))==0 and np.max(np.abs(y))==0:\n",
    "            continue\n",
    "        else:\n",
    "            log2_mfr = np.log2(np.mean(x)/np.mean(y))\n",
    "            pvalue = stats.ks_2samp(x, y)[1]\n",
    "            flux_res.append([cond1, cond2, rid, np.mean(x), np.mean(y), log2_mfr, pvalue])\n",
    "df_flux_res = pd.DataFrame(flux_res, columns=['Condition1','Condition2','Reaction','MeanFluxCond1','MeanFluxCond2','Log2MeanFluxRatio','P'])\n",
    "df_flux_res['Padj'] = multipletests(df_flux_res.P, alpha=0.05, method='fdr_bh')[1]\n",
    "df_flux_res.to_csv(\"flux_sampling_comparison_summary.csv\", index=False)\n",
    "df_flux_res.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 291,
   "id": "468865d8-ff73-4c9e-88b3-8069af72a001",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Condition1</th>\n",
       "      <th>Condition2</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>MeanFluxCond1</th>\n",
       "      <th>MeanFluxCond2</th>\n",
       "      <th>Log2MeanFluxRatio</th>\n",
       "      <th>P</th>\n",
       "      <th>Padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3291</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>GD1Atg</td>\n",
       "      <td>0.055082</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1754</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>RE2994X</td>\n",
       "      <td>0.054533</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3865</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>HMR_3346</td>\n",
       "      <td>0.054520</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3864</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>HMR_3345</td>\n",
       "      <td>0.054520</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3863</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>HMR_3344</td>\n",
       "      <td>0.054520</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3461</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>HMR_0823</td>\n",
       "      <td>0.051022</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3862</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>HMR_3343</td>\n",
       "      <td>0.054520</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3468</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>HMR_0837</td>\n",
       "      <td>0.053706</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3861</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>HMR_3342</td>\n",
       "      <td>0.054520</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1756</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>RE2997X</td>\n",
       "      <td>0.054533</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1755</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>RE2996X</td>\n",
       "      <td>0.054533</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3860</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>HMR_3341</td>\n",
       "      <td>0.054520</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3859</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>HMR_3340</td>\n",
       "      <td>0.054520</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1753</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>RE2993X</td>\n",
       "      <td>0.054533</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1822</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>RE3234C</td>\n",
       "      <td>0.054520</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1752</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>RE2992X</td>\n",
       "      <td>0.054533</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3858</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>HMR_3339</td>\n",
       "      <td>0.054520</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3857</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>HMR_3338</td>\n",
       "      <td>0.054520</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1750</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>RE2991X</td>\n",
       "      <td>0.054533</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1749</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>RE2990X</td>\n",
       "      <td>0.054533</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1748</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>RE2989X</td>\n",
       "      <td>0.054533</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3856</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>HMR_3337</td>\n",
       "      <td>0.054520</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1747</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>RE2988X</td>\n",
       "      <td>0.054533</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1746</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>RE2987X</td>\n",
       "      <td>0.054533</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1745</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>RE2986X</td>\n",
       "      <td>0.054533</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1744</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>RE2985X</td>\n",
       "      <td>0.054533</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1815</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>RE3226C</td>\n",
       "      <td>0.054520</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1823</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>RE3235C</td>\n",
       "      <td>0.054520</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>B3GALT42g</td>\n",
       "      <td>0.055082</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1938</th>\n",
       "      <td>naive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>CHOLESTle_1</td>\n",
       "      <td>0.052112</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Condition1 Condition2     Reaction  MeanFluxCond1  MeanFluxCond2  \\\n",
       "3291      naive   wildtype       GD1Atg       0.055082            0.0   \n",
       "1754      naive   wildtype      RE2994X       0.054533            0.0   \n",
       "3865      naive   wildtype     HMR_3346       0.054520            0.0   \n",
       "3864      naive   wildtype     HMR_3345       0.054520            0.0   \n",
       "3863      naive   wildtype     HMR_3344       0.054520            0.0   \n",
       "3461      naive   wildtype     HMR_0823       0.051022            0.0   \n",
       "3862      naive   wildtype     HMR_3343       0.054520            0.0   \n",
       "3468      naive   wildtype     HMR_0837       0.053706            0.0   \n",
       "3861      naive   wildtype     HMR_3342       0.054520            0.0   \n",
       "1756      naive   wildtype      RE2997X       0.054533            0.0   \n",
       "1755      naive   wildtype      RE2996X       0.054533            0.0   \n",
       "3860      naive   wildtype     HMR_3341       0.054520            0.0   \n",
       "3859      naive   wildtype     HMR_3340       0.054520            0.0   \n",
       "1753      naive   wildtype      RE2993X       0.054533            0.0   \n",
       "1822      naive   wildtype      RE3234C       0.054520            0.0   \n",
       "1752      naive   wildtype      RE2992X       0.054533            0.0   \n",
       "3858      naive   wildtype     HMR_3339       0.054520            0.0   \n",
       "3857      naive   wildtype     HMR_3338       0.054520            0.0   \n",
       "1750      naive   wildtype      RE2991X       0.054533            0.0   \n",
       "1749      naive   wildtype      RE2990X       0.054533            0.0   \n",
       "1748      naive   wildtype      RE2989X       0.054533            0.0   \n",
       "3856      naive   wildtype     HMR_3337       0.054520            0.0   \n",
       "1747      naive   wildtype      RE2988X       0.054533            0.0   \n",
       "1746      naive   wildtype      RE2987X       0.054533            0.0   \n",
       "1745      naive   wildtype      RE2986X       0.054533            0.0   \n",
       "1744      naive   wildtype      RE2985X       0.054533            0.0   \n",
       "1815      naive   wildtype      RE3226C       0.054520            0.0   \n",
       "1823      naive   wildtype      RE3235C       0.054520            0.0   \n",
       "132       naive   wildtype    B3GALT42g       0.055082            0.0   \n",
       "1938      naive   wildtype  CHOLESTle_1       0.052112            0.0   \n",
       "\n",
       "      Log2MeanFluxRatio    P  Padj  \n",
       "3291                inf  0.0   0.0  \n",
       "1754                inf  0.0   0.0  \n",
       "3865                inf  0.0   0.0  \n",
       "3864                inf  0.0   0.0  \n",
       "3863                inf  0.0   0.0  \n",
       "3461                inf  0.0   0.0  \n",
       "3862                inf  0.0   0.0  \n",
       "3468                inf  0.0   0.0  \n",
       "3861                inf  0.0   0.0  \n",
       "1756                inf  0.0   0.0  \n",
       "1755                inf  0.0   0.0  \n",
       "3860                inf  0.0   0.0  \n",
       "3859                inf  0.0   0.0  \n",
       "1753                inf  0.0   0.0  \n",
       "1822                inf  0.0   0.0  \n",
       "1752                inf  0.0   0.0  \n",
       "3858                inf  0.0   0.0  \n",
       "3857                inf  0.0   0.0  \n",
       "1750                inf  0.0   0.0  \n",
       "1749                inf  0.0   0.0  \n",
       "1748                inf  0.0   0.0  \n",
       "3856                inf  0.0   0.0  \n",
       "1747                inf  0.0   0.0  \n",
       "1746                inf  0.0   0.0  \n",
       "1745                inf  0.0   0.0  \n",
       "1744                inf  0.0   0.0  \n",
       "1815                inf  0.0   0.0  \n",
       "1823                inf  0.0   0.0  \n",
       "132                 inf  0.0   0.0  \n",
       "1938                inf  0.0   0.0  "
      ]
     },
     "execution_count": 291,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_flux_res[(df_flux_res.Condition1=='naive') & (df_flux_res.Condition2=='wildtype')].sort_values('Log2MeanFluxRatio', ascending=False).head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "id": "750bd3bd-5853-41f8-8429-48e38d9de0fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>EX_h2o2_e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Hydrogen peroxide exchange</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01399c2590</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>h2o2_e --></p>\n",
       "                    <p style='text-align:right'>Hydrogen Peroxide --></p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction EX_h2o2_e at 0x1399c2590>"
      ]
     },
     "execution_count": 296,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unconstrained_model.reactions.EX_h2o2_e"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 298,
   "id": "16bd47f4-8420-412c-9e77-a1fd5bc6e1eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Condition1</th>\n",
       "      <th>Condition2</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>MeanFluxCond1</th>\n",
       "      <th>MeanFluxCond2</th>\n",
       "      <th>Log2MeanFluxRatio</th>\n",
       "      <th>P</th>\n",
       "      <th>Padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Condition1, Condition2, Reaction, MeanFluxCond1, MeanFluxCond2, Log2MeanFluxRatio, P, Padj]\n",
       "Index: []"
      ]
     },
     "execution_count": 298,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_flux_res[df_flux_res.Reaction==\"EX_h2o2_e\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 303,
   "id": "c775b1a3-eea6-4105-b573-8529016a176d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0"
      ]
     },
     "execution_count": 303,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flux_samples['naive'][\"EX_h2o2_e\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22d7fd94-659b-4121-b854-3d663784eba3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
